### PR TITLE
Ginkgo: Call AfterAll if it fails on BeforeAll

### PR DESCRIPTION
**Pull request by eloycoto**
@tgraf discovered that if a BeforeAll trigger an assertion the AfterAll
code was not executed so the machine was not cleaned correctly.

With this change the counters for AfterAll are triggered on RunAfterEach
where all the functions happens and does not matter if the IT is
executed or not.

Due the debugging also found that the order for AfterEach was not
correct in terms of AfterAll in a context. Fix that issue also.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6598)

### CODE REVIEW

**Comment by eloycoto** [comment]
@ianvernon I know that you're going to ask: 
```
package RuntimeTest

import (
	"fmt"

	. "github.com/cilium/cilium/test/ginkgo-ext"

	. "github.com/onsi/gomega"
)

var _ = Describe("RuntimeEloy", func() {

	BeforeAll(func() {
		fmt.Println("->BeforeAll")
	})

	BeforeEach(func() {
		fmt.Println("-->BeforeEach")

	})
	JustAfterEach(func() {
		fmt.Println("->JustAfterEach")
	})

	AfterFailed(func() {
		fmt.Println("-->AfterFailed")
	})

	AfterEach(func() {
		fmt.Println("-->AfterEach")
	})

	AfterAll(func() {
		fmt.Println("--->AfterAll")
	})

	It("Simple test", func() {
		fmt.Println("Simple test")
	})

	Context("New Context", func() {

		BeforeAll(func() {
			fmt.Println("-#:BeforeAll")
			Expect(true).To(BeFalse())
		})

		BeforeEach(func() {
			fmt.Println("--#:BeforeEach")
			Expect(false).To(BeTrue())
		})

		JustAfterEach(func() {
			fmt.Println("-#:JustAfterEach")
		})

		AfterFailed(func() {
			fmt.Println("--#:AfterFailed")
		})

		AfterEach(func() {
			fmt.Println("--#:AfterEach")
		})

		AfterAll(func() {
			fmt.Println("---#:AfterAll")
		})

		It("Simple ctx test", func() {
			fmt.Println("Simple test")
		})

		It("Simple ctx test b", func() {
			fmt.Println("Simple test")
		})
	})

	It("Simple testB", func() {
		fmt.Println("Simple test")
	})
})

```

```

breo.acalustra.com ➜  test git:(master) ✗ ginkgo --focus "RuntimeEloy" -v -- --cilium.provision=false
level=info msg="environment variable \"K8S_VERSION\" was not set; setting to default value \"1.13\""
Running Suite: runtime
======================
Random Seed: 1547129935
Will run 4 of 327 specs

S
------------------------------
RuntimeEloy
  Simple test
  /home/eloy/.go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:408
->BeforeAll
-->BeforeEach
Simple test
=== Test Finished at 2019-01-10T15:18:58+01:00====
->JustAfterEach
-->AfterEach
<Checks>

</Checks>

•
------------------------------
RuntimeEloy New Context
  Simple ctx test
  /home/eloy/.go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:408
-->BeforeEach
-#:BeforeAll
-#:JustAfterEach
->JustAfterEach
===================== TEST FAILED =====================
--#:AfterFailed
===================== Existing AfterFailed =====================
===================== TEST FAILED =====================
-->AfterFailed
===================== Existing AfterFailed =====================
--#:AfterEach
-->AfterEach
<Checks>

</Checks>


• Failure in Spec Setup (BeforeEach) [0.001 seconds]
RuntimeEloy
/home/eloy/.go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:376
  New Context
  /home/eloy/.go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:376
    Simple ctx test [BeforeEach]
    /home/eloy/.go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:408

    Expected
        <bool>: true
    to be false

    /home/eloy/.go/src/github.com/cilium/cilium/test/runtime/eloy.go:45
------------------------------
RuntimeEloy New Context
  Simple ctx test b
  /home/eloy/.go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:408
-->BeforeEach
-#:JustAfterEach
->JustAfterEach
===================== TEST FAILED =====================
--#:AfterFailed
===================== Existing AfterFailed =====================
===================== TEST FAILED =====================
-->AfterFailed
===================== Existing AfterFailed =====================
--#:AfterEach
-->AfterEach
<Checks>

</Checks>

---#:AfterAll

• Failure in Spec Setup (BeforeEach) [0.001 seconds]
RuntimeEloy
/home/eloy/.go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:376
  New Context
  /home/eloy/.go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:376
    Simple ctx test b [BeforeEach]
    /home/eloy/.go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:408

    failed due to BeforeAll failure

    /home/eloy/.go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:491
------------------------------
RuntimeEloy
  Simple testB
  /home/eloy/.go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:408
-->BeforeEach
Simple test
=== Test Finished at 2019-01-10T15:18:58+01:00====
->JustAfterEach
-->AfterEach
<Checks>

</Checks>

--->AfterAll
•SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSAfterSuite: not running on Jenkins; leaving VMs running for debugging


Summarizing 2 Failures:

[Fail] RuntimeEloy New Context [BeforeEach] Simple ctx test
/home/eloy/.go/src/github.com/cilium/cilium/test/runtime/eloy.go:45

[Fail] RuntimeEloy New Context [BeforeEach] Simple ctx test b
/home/eloy/.go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:491

Ran 4 of 327 Specs in 0.005 seconds
FAIL! -- 2 Passed | 2 Failed | 0 Pending | 323 Skipped --- FAIL: TestTest (0.01s)
FAIL

Ginkgo ran 1 suite in 3.64844981s
Test Suite Failed
```

**Comment by eloycoto** [comment]
test-me-please

**Comment by ianvernon** [comment]
@eloycoto thank you for providing example output. In the first failure above:

```
RuntimeEloy New Context
  Simple ctx test
  /home/eloy/.go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:408
-->BeforeEach
-#:BeforeAll
-#:JustAfterEach
->JustAfterEach
===================== TEST FAILED =====================
--#:AfterFailed
===================== Existing AfterFailed =====================
===================== TEST FAILED =====================
-->AfterFailed
===================== Existing AfterFailed =====================
--#:AfterEach
-->AfterEach
<Checks>

</Checks>


• Failure in Spec Setup (BeforeEach) [0.001 seconds]
RuntimeEloy
/home/eloy/.go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:376
  New Context
  /home/eloy/.go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:376
    Simple ctx test [BeforeEach]
    /home/eloy/.go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:408

    Expected
        <bool>: true
    to be false

    /home/eloy/.go/src/github.com/cilium/cilium/test/runtime/eloy.go:45
```

Is there a reason it says `Failure in Spec Setup (BeforeEach)` - shouldn't this say `Failure in Spec Setup (BeforeAll)`? I thought that the failure would occur in the `BeforeAll`?

The rest makes sense to me in that it shows that the `AfterAll` is ran after failures occurred in `BeforeEach` and `BeforeAll` within the `Context`.

> Due the debugging also found that the order for AfterEach was not
correct in terms of AfterAll in a context. Fix that issue also.

What was incorrect about the order exactly before compared to the changes made here? Does this mean the `AfterEach` that is contained within the `Context` is ran, then the `AfterEach` from outside the `Context` within the `Describe` is ran, before the `AfterAll` within the `Context`? I disagree with that implementation. The point of a `Context` is to wrap certain behavior within a block within a `Describe`, so I think that the all of the items within a `Context` should run before anything outside of the `Context` is ran? E.g., `AfterEach` in Context --> `AfterAll` in Context --> `AfterEach` outside of `Context`.

File: test/ginkgo-ext/scopes.go [hunk 1]
```diff
@@ -216,6 +216,27 @@ func runAllJustAfterEach(cs *scope, testName string) {
 	}
 }
 
+// afterEachStatus map to store what `AfterEach` functions have been
+// already executed for the given test
+var afterEachStatus map[string]bool = map[string]bool{}
+
+// runAllAfterEach runs all the `scope.AfterEach` functions for the
+// given scope and parent scopes. This function make sure that all the
+// `AfterEach` functions are called before AfterAll functions.
```
Comment (ianvernon) [inline test/ginkgo-ext/scopes.go:225]: In line with my comment on this PR - what is the rationale behind this change?

**Comment by eloycoto** [comment]
> Is there a reason it says Failure in Spec Setup (BeforeEach) - shouldn't this say Failure in Spec Setup (BeforeAll)? I thought that the failure would occur in the BeforeAll?

We have "support" over AfterEach, but the beforeEach we don't have too much control. So the BeforeAll is only a BeforeEach that is taking care of happen first. 

>What was incorrect about the order exactly before compared to the changes made here? Does this mean the AfterEach that is contained within the Context is ran, then the AfterEach from outside the Context within the Describe is ran, before the AfterAll within the Context? I disagree with that implementation. The point of a Context is to wrap certain behavior within a block within a Describe, so I think that the all of the items within a Context should run before anything outside of the Context is ran? E.g., AfterEach in Context --> AfterAll in Context --> AfterEach outside of Context.

The thing about context, no it's not like that, the AfterEach is executed all. If you want to have your behaviour need to be in another Describe function.  Before this change the behaviour is: 
```
Describe
AfterEach(a)
AfterAll(a)
Context
    AfterEach(b)
    AfterAll(b)
    test()
```

The execution will be:
```
test
AfterEach(b)
AfterAll(b)
AfterEach(a)
AfterAll(a)
```

But with the changes the behaviour will be: 

```
test
AfterEach(b)
AfterEach(a)
AfterAll(b)
AfterAll(a)
```

Your comment does not make too much sense to me, due if you do that, we need to
define a AfterFail, and JustAfterEach in each block, if not the data will not
work(In the past we made this change to support this)

Now the execution is like the docs: 
http://docs.cilium.io/en/v1.3/contributing/#example-test-layout

**Comment by ianvernon** [comment]
@eloycoto the main problem I have with the approach you've suggested is that it seems to violate the point of what a `Context` does, which is to sort-of 'namespace' a set of tests within a `Describe`. 

For instance, looking at the following which you proposed:

```
test
AfterEach(b)
AfterEach(a)
AfterAll(b)
AfterAll(a)
```

I don't see why something declared outside of the `Context` (`AfterEach(a)`) should run before the `AfterAll(b)` which is declared inside the `Context`. I think this is similar to what we disagreed on with `AfterFailed`. I don't understand what makes the order you propose correct, due to the statements I have made above. Is it to reduce the amount of boilerplate code a-la our discussion around `AfterFailed`?

>  if not the data will not
work(In the past we made this change to support this)

Can you elaborate on what "the data" means? I don't understand. Are you referring to the running of `AfterFailed` within the `Context` even if the `AfterFailed` is declared outside of the `Context`? 

While I disagree per my reasons above, I don't want to be too fussy. Feel free to merge this since this is in-line with what the documentation states.

**Comment by eloycoto** [comment]
@ianvernon thanks for the feedback! I'll merge this one and let's discuss the behaviour in the near future, I do not agree with you, but at the moment some builds are not working as the doc, so I think that need to be merged. 

Let's take this offline and make an RFC for the change. 

Regards.
